hazelcast/src/test/java/com/hazelcast/test/ParallelParameterized.java

```diff
@@ -16,16 +16,13 @@
 
 package com.hazelcast.test;
 
-/**
- * Marker thread subclass for threads created by {@link HazelcastParallelClassRunner}.
- */
-class MultithreadedTestRunnerThread extends Thread {
+import org.junit.runners.Parameterized;
 
-    MultithreadedTestRunnerThread(Runnable target) {
-        super(target);
-    }
+public class ParallelParameterized extends Parameterized {
+
+    public ParallelParameterized(Class<?> klass) throws Throwable {
+        super(klass);
 
-    MultithreadedTestRunnerThread(Runnable target, String name) {
-        super(target, name);
+        setScheduler(new HzRunnerScheduler());
```

Comment:
sorry, didn't get this. can you please give a concrete example?

Reply:
If I get the situation right, the situation is the following. The serial runner allows only a single test case to run at a time, while the parallel one allows multiple cases in parallel. Then the parameterized tests sit on top of this. My understanding is that
1) this PR allows running the parameterized test cases in parallel regardless of which runner is being used, which I think conflicts with the contract of the serial runner a bit;
2) on the efficiency side, the serial runner will take the next test case only if all parameterized versions of the currently and concurrently running test case complete, which will mean there will be idle threads in the pool of the scheduler's executor before moving on to the next test case (assuming the parallel parameterized cases complete at different times). 

I'm asking if there is any benefit in this approach compared to using the old approach for making the parameterized tests running in parallel (so relying on the parallel runner), but with this new scheduler, which looks better than the old one with the time-based backoff logic.

Not a concrete example, but I hope my point is a bit clearer now :smile:

Reply:
Let me put some examples here.

This configuration runs tests serially, one by one, like before:
```
@RunWith(Parameterized.class)
@UseParametersRunnerFactory(HazelcastSerialParametersRunnerFactory.class)
```

This configuration runs tests in parallel like before:
```
@RunWith(Parameterized.class)
@UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
```

This PR adds an additional option--> `@RunWith(ParallelParameterized.class)`
If you run with it, parallelism will increase. Because it parallelize all possible test combinations, this is an enhancement over what we have now, because current parallelism happens per parameter value. Means, if there are 3 possible values for a parameter and 3 test methods in the class. Current runner parallelize 3 tests per value, but new runner parallelize 3x3 = 9 tests.